[SPARK-27258][K8S]Deal with the k8s resource names that don't match their own regular expression

hehuiyuan opened the pull request:
## What changes were proposed in this pull request?



The regex of Pod is `^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*`.
 The regex of Servive is `^[a-z]([-a-z0-9]*[a-z0-9])?`.
The regex of Secret is `^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*`.

If --name=-min, the first character is '-' ,which does not satisfy the regex for Pod ,Service,Secret.
if --name=1min, the first character is digit, which does not satisfy the regex for Service.

Commit:
Fix the first character of  Service's name is digit

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverServiceFeatureStep.scala:
```diff
@@ -38,15 +38,21 @@ private[spark] class DriverServiceFeatureStep(
   private val preferredServiceName = s"${kubernetesConf.resourceNamePrefix}$DRIVER_SVC_POSTFIX"
-  private val resolvedServiceName = if (preferredServiceName.length <= MAX_SERVICE_NAME_LENGTH) {
+  private val resolvedServiceName = if (preferredServiceName.length <= MAX_SERVICE_NAME_LENGTH
+    && !isDigit(preferredServiceName.charAt(0))) {
```
This is still wrong. You are saying the first character needs to be a-z. This only checks if it's a digit.

hehuiyuan replied:
!isDigit(preferredServiceName.charAt(0))   ：
if true :  that means the first character is not digit；
if false : That means the first character is digit.

If the first character is digit ,  see the code `else{...}`.
If the first character is not digit, we don't need to deal with the name of Service.

srowen replied:
You're not seeing my point. I understand what this code does.
The title of this PR says that you must check if the first character is a-z. That is not the same as being a digit.

hehuiyuan replied:
I want to express :
The name of Service needs to satisfies the regular expression. Otherwise, we need to deal with it further.

srowen replied:
Look at the first part of the regular expression you are trying to match. It does not say "not a digit" which is what you are checking. You need to check whether it's a-z.

hehuiyuan replied:
I got it. You mean  the first character is not only digit such as other symbols.

hehuiyuan replied:
I say digit,  because resourceNamePrefix  that has been processed and contains three kinds :
(1)digit
(2)a-z
(3)- 


```
  def getResourceNamePrefix(appName: String): String = {
    val id = KubernetesUtils.uniqueID()
    s"$appName-$id"
      .trim
      .toLowerCase(Locale.ROOT)
      .replaceAll("\\s+", "-")
      .replaceAll("\\.", "-")
      .replaceAll("[^a-z0-9\\-]", "")
      .replaceAll("-+", "-")
  }
```

srowen replied:
Sure, but why not make the logic consistent with the desired outcome here? it's just simpler and catches the '-' case.

hehuiyuan replied:
I think only Service name needs  to satisfy that the first character is a-z. The Pod name can pass the validation for pod's regex when use resourceNamePrefix.

srowen replied:
This change is still wrong then. You are not checking whether the first char is a-z.

hehuiyuan replied:
You are right. I add a regex `replaceAll("^-", "")`.
Now ,the method `getResourceNamePrefix ` returns  value that the first character is 0-9 or a-z.

srowen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverServiceFeatureStep.scala:
```diff
@@ -38,15 +38,21 @@ private[spark] class DriverServiceFeatureStep(
     s"$DRIVER_HOST_KEY is not supported in Kubernetes mode, as the driver's hostname will be " +
       "managed via a Kubernetes service.")
 
+  def isDigit(char: Char) : Boolean = {
```
There's already a JDK method for this anyway, but this isn't the check you want to make.

hehuiyuan replied:
Character.isDigit(preferredServiceName.charAt(0))

Commit:
use JDK method 'Character.isDigit'

hehuiyuan commented:
I want to express :
  The name of Service  needs to satisfies the regular expression. Otherwise, we need to deal with it further.

Commit:
add extra regex for getResourceNamePrefix method

vanzin commented:
Instead of adding more ad-hoc tests, I think it's better to just check if the preferred name is valid using `com.google.common.net.InternetDomainName.isValid`. It seems to do a better job than the checks that were put in this code originally.

srowen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala:
```diff
@@ -198,5 +198,6 @@ private[spark] object KubernetesConf {
       .replaceAll("\\.", "-")
       .replaceAll("[^a-z0-9\\-]", "")
       .replaceAll("-+", "-")
+      .replaceAll("^-", "")
```
Why? this character is allowed later in the name. I don't get why you can't just check a-z? I don't think this PR is making progress, and will close it

hehuiyuan replied:
I replaced the first character，ok?

hehuiyuan replied:
I updated the code for checking a-z.
`Character.isLetter(preferredServiceName.charAt(0)`

hehuiyuan replied:
`.replaceAll("^-", "")` ,we can use the code to avoid that the first character of all kubernetes resources(Pod ,Service,Secret,etc) name is not '-'

srowen replied:
This should be reverted then?

Commit:
replace isDigt with isLetter

hehuiyuan commented:
> Instead of adding more ad-hoc tests, I think it's better to just check if the preferred name is valid using `com.google.common.net.InternetDomainName.isValid`. It seems to do a better job than the checks that were put in this code originally.

`.replaceAll("^-", "")` ,we can use the code to avoid that the first character of all kubernetes resources(Pod ,Service,Secret,etc) name is '-'.
Then `Character.isLetter(preferredServiceName.charAt(0)` is used to checking the Service name

SparkQA commented:
**[Test build #4671 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4671/testReport)** for PR 24219 at commit [`6af6557`](https://github.com/apache/spark/commit/6af655706290c1416342a3ba191ae28639115119).

SparkQA commented:
**[Test build #4671 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4671/testReport)** for PR 24219 at commit [`6af6557`](https://github.com/apache/spark/commit/6af655706290c1416342a3ba191ae28639115119).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

skonto commented:
There are some utility methods that could be used offered by the fabric8io client for this purprose:
https://github.com/fabric8io/kubernetes-client/blob/v4.2.2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java
For example `sanitizeName` method etc.

hehuiyuan commented:
> There are some public utility methods that could be used offered by the fabric8io client for this purprose:
> https://github.com/fabric8io/kubernetes-client/blob/v4.2.2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java
> For example `io.fabric8.kubernetes.client.utils.KubernetesResourceUtil.{sanitizeName(),isValidName()}` methods etc.
> This way code is cleaner and we dont need to maintain changes as K8s evolves.

Hi, that sounds good. This class supports many methods.  But  I don't found the method that processes the first character of resourceName is `-` and  checks the name of Service in KubernetesResourceUtil .

Service  is different from the others.

```
The regex of Servive is ^[a-z]([-a-z0-9]*[a-z0-9])?.
The regex of Pod is ^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*.
The regex of Secret is ^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*.
```

skonto commented:
`sanitize` replaces "-".

skonto commented:
`isValidName()` method uses "[a-z0-9]([-a-z0-9]*[a-z0-9])?" which is in compliance with the[ k8s codebase](https://github.com/kubernetes/kubernetes/blob/6902f3112d98eb6bd0894886ff9cd3fbd03a7f79/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L110) for DNS-1123 labels and works for service names. 
Sanitize does not fix the dash issue as a first character I agree, btw this happened before because of [truncation](https://github.com/apache/spark/pull/23781), but the sanitize method could be improved or re-use the logic for the executor pod names [here](https://github.com/apache/spark/blob/master/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala#L87).
So its not just dash. I would refactor the code and use only one method to sanitize `resourceNamePrefix`  because `resourceNamePrefix` can be used by the executor pods for the hostname thing anyway.

hehuiyuan commented:
> `sanitize()` method does not fix the dash issue as a first character I agree, btw this happened before because of [truncation](https://github.com/apache/spark/pull/23781), but the sanitize method could be improved or re-use the logic for the executor pod names [here](https://github.com/apache/spark/blob/master/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala#L87) and create one re-usable method. Pod names for example can be set in different ways eg via `spark.kubernetes.driver.pod.name`, scenario currently not covered. Code should be refactored isolating the logic in new methods like getPodName, getServiceName and so on, so validation and sanitization/truncation can happen in one place for the final string. I general I dont think these details should be managed by user code, they should be offered by the client (we could create a PR against fabric8io if makes sense but anyway).

That sounds good.

hehuiyuan commented:
service name is a-z ，which is not a-z0-9


 
---Original---
From: "Stavros Kontopoulos"<notifications@github.com>
Date: Tue, May 14, 2019 17:06 PM
To: "apache/spark"<spark@noreply.github.com>;
Cc: "Author"<author@noreply.github.com>;"hehuiyuan"<471627698@qq.com>;
Subject: Re: [apache/spark] [SPARK-27258][K8S]Deal with the k8s resource names that don't match their own regular expression  (#24219)



isValidName() method uses "a-z0-9?" which is in compliance with the k8s codebase for DNS-1123 labels and works for service names.
 Sanitize does not fix the dash issue as a first character I agree, btw this happened before because of truncation, but the sanitize method could be improved or re-use the logic for the executor pod names here.
 So its not just dash. I would refactor the code and use only one method to sanitize resourceNamePrefix  because resourceNamePrefix can be used by the executor pods for the hostname thing anyway.
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

skonto commented:
@hehuiyuan I removed that comment :)

hehuiyuan commented:
> @hehuiyuan I removed that comment :) , yes indeed no digits.

Hi, I have extracted the methods to get the name of the  resources for Service / Pod / Secret / ConfigMap etc.

skonto commented:
@hehuiyuan any progress on this one?

hehuiyuan commented:
> @hehuiyuan any progress on this one?

![resourcename](https://user-images.githubusercontent.com/18002496/58886128-f8381a00-8715-11e9-8b64-f7e688d5a0c1.jpg)


Is this your meaning ?

skonto commented:
I mean are all the comments addressed? I think vanzin has one above. @srowen should this be merged?

srowen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverServiceFeatureStep.scala:
```diff
@@ -39,14 +39,16 @@ private[spark] class DriverServiceFeatureStep(
-      s"too long (must be <= $MAX_SERVICE_NAME_LENGTH characters). Falling back to use " +
-      s"$shorterServiceName as the driver service's name.")
+      s"too long (must be <= $MAX_SERVICE_NAME_LENGTH characters) " +
+      s"or the first character of $preferredServiceName is not letter which is not support." +
```
support -> supported.
You could just say "or is not a valid service name"

hehuiyuan replied:
OK! But there are some conflicts at present.

hehuiyuan replied:
@srowen Hi, i have a  question  that is how to deal with the prefix of user-defined pod.

It handles  the default prefix of podName in this PR .
Is there any need for further treatment or it's necessary to explain it in the doc.

```scala
  override val resourceNamePrefix: String = {
    get(KUBERNETES_EXECUTOR_POD_NAME_PREFIX).getOrElse(
      KubernetesConf.getResourceNamePrefix(appName))
  }

  private val driverPodName = conf
    .get(KUBERNETES_DRIVER_POD_NAME)
    .getOrElse(s"${conf.resourceNamePrefix}-driver")

```

srowen replied:
That I don't know ... @skonto?
